### PR TITLE
feat: Add dependencies field to `ComputedFieldInfo`

### LINKED ISSUE #9893: Add "dependencies" field to `ComputedFieldInfo`
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

It would be nice to have a way to indicate that a given `computed_field` depends on the presence/values of other fields.  This could be used, for example, to indicate that a given computed (`readOnly`) field is likely to be stale if another field (on which it depends) has changed.  I'm not proposing that pydantic do anything with this info at the moment; just allow model creators to make it available to consumers in `ComputedFieldInfo`. 

This is also related (albeit somewhat indirectly) to the [`dependentRequired` concept in JSON Schema](https://json-schema.org/understanding-json-schema/reference/conditionals#dependentRequired) ... which allows you to specify that the presence of one field requires the presence of another field.

Taking the computed_field example in the docs, I'm proposing:

```python
class Box(BaseModel):
    width: float
    height: float = 1
    depth: float = 2

    @computed_field(dependencies=['width', 'height', 'depth'])
    def volume(self) -> float:
        return self.width * self.height * self.depth
```

in terms of a serialization schema, this would be similar to

```json
{
  "title": "Box",
  "type": "object",
  "properties": {
    "width": { "title": "Width", "type": "number" },
    "height": { "default": 1, "title": "Height", "type": "number" },
    "depth": { "default": 2, "title": "Depth", "type": "number" },
    "volume": { "readOnly": true, "title": "Volume", "type": "number" }
  },
  "required": ["width"],
  "dependentRequired": {
    "volume": ["width", "height", "depth"]
  }
}
```

### Alternative

Currently, if I'd like to know what fields a computed field might depend on, I'm mostly left to very error prone parsing of source code?

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [X] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by tlambert03**
## Change Summary

Allows the model creator to give a hint as to which other fields a `computed_field` depends on:

```python
    class Model(BaseModel):
        width: float
        height: float

        @computed_field(dependencies=['width', 'height'])
        @property
        def area(self) -> float:
            return self.width * self.height
```

## Related issue number

closes #9893

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by tlambert03** [comment]
cc @sydney-runkle, this really was all i was proposing for now.  I stopped short of adding [dependentRequired](https://json-schema.org/understanding-json-schema/reference/conditionals#dependentRequired) to the JSON schema output, since i'm not sure it's a perfect semantic match.  but I could add that if you or others would like

**Comment by Viicos** [comment]
> I stopped short of adding [dependentRequired](https://json-schema.org/understanding-json-schema/reference/conditionals#dependentRequired) to the JSON schema output, since i'm not sure it's a perfect semantic match

Indeed seems like `dependentRequired` makes more sense a a validation schema (and `computed_field`s are only relevant during serialization), as in "if _foo_ is provided, _bar_ should be provided as well). It could make sense during serialization as well ("if _foo_ is present, _bar_ is guaranteed to be there as well), although this does not really apply here, as I don't think there's a native way with Pydantic to express this conditional presence of a field during serialization.

File: tests/test_computed_fields.py [hunk 1]
```diff
@@ -795,3 +795,16 @@ def id(self) -> str:
...
     m = Model(bar=42)
     assert m.model_dump() == {'bar': 42, 'id': 'id: {"bar":42}'}
+
+
+def test_computed_field_dependencies():
+    class Model(BaseModel):
+        width: float
+        height: float
+
+        @computed_field(dependencies=['width', 'height'])
```
Comment (Viicos) [inline tests/test_computed_fields.py:805]: How should nested fields be specified? e.g.:

```python
class Sub(BaseModel):
    foo: int

class Model(BaseModel):
    sub: Sub

    @computed_field(dependencies=[???])
    @property
    def prop(self) -> int:
        return self.sub.int
```

Comment (tlambert03) [reply tests/test_computed_fields.py:805]: ah, very good question. 
i guess the two possibilities that come to mind are:

- follow the same `loc` syntax that pydantic uses when expressing Validation errors, with dotted attributes  `dependencies=['sub.foo']`
- allow deps to be a dict as well `dependencies={'sub': ['foo']}`
- or... just say "you can't": *all* names in dependencies must be field names on this model `dependencies=['sub']`

I admit that None of those make me jump for joy, and that does definitely raise an issue for the whole pattern.  What do you think?

Comment (sydney-runkle) [reply tests/test_computed_fields.py:805]: I'd be inclined to use a pattern similar to that of `exclude` as a serialization flag:

https://docs.pydantic.dev/latest/concepts/serialization/#advanced-include-and-exclude

File: pydantic/fields.py [hunk 2]
```diff
@@ -1027,6 +1028,7 @@ class ComputedFieldInfo:
     examples: list[Any] | None
     json_schema_extra: JsonDict | typing.Callable[[JsonDict], None] | None
     repr: bool
+    dependencies: list[str]
```
Comment (sydney-runkle) [inline pydantic/fields.py:1031]: Why not `Sequence[str]` here? I suppose we're probably about to change this for nested values anyways...

File: pydantic/fields.py [hunk 3]
```diff
@@ -1013,6 +1013,7 @@ class ComputedFieldInfo:
         examples: Example values of the computed field to include in the serialization JSON schema.
         json_schema_extra: A dict or callable to provide extra JSON schema properties.
         repr: A boolean indicating whether to include the field in the __repr__ output.
+        dependencies: A list of field names that the computed field depends on for its value.
```
Comment (sydney-runkle) [inline pydantic/fields.py:1016]: Let's add the note about json schema here as well

**Comment by sydney-runkle** [comment]
Tentatively looks good to me, but I have a feeling that folks will want to extend the functionality of this beyond just JSON schema. This is a great place to start though!

File: pydantic/fields.py (hunk 2 above)
Comment (tlambert03) [reply pydantic/fields.py:1031]: yeah, will change soon anyway, as per your suggestion.  But here, i had used `Sequence[str]` for what we accept from the user in the `computed_field` decorator... but cast that all to `list[str]` for uniformity on this internal object ("Postel's law" style)

**Comment by Viicos** [comment]
@sydney-runkle note that this PR does not add anything related to JSON Schema yet, and as discussed in the previous comments `dependentRequired` isn't truly related

**Comment by sydney-runkle** [comment]
Going to chat with the team about this at our oss meeting tomorrow - will follow up after that!

**Comment by sydney-runkle** [comment]
Alright, after thinking more about this, I think @Viicos has a good point here: https://github.com/pydantic/pydantic/issues/9893#issuecomment-2228091121

> So perhaps a metadata generic argument, typed as dict[str, Any]. I'll note that this isn't specific to computed_field, it could be used for normal fields as well.

I think it'd make sense to take this approach here and also sets us up well for future requests like this one.

I'd be more inclined to introduce `dependencies` independently if we reflected said info directly in the JSON schema, or if we could enforce the dependency. @adriangb had a good point here about the difficulties of that approach: https://github.com/pydantic/pydantic/issues/6085#issuecomment-1599092334. Basically, enforcing dependencies between fields is difficult, and not something we're planning on doing.

**Comment by tlambert03** [comment]
> I think it'd make sense to take this approach here and also sets us up well for future requests like this one.

that approach was basically to leave it up to end users to come up with their own conventions for expressing what fields a computed field may depend on, correct?  In other words, to make a generic, unstructured key value field for arbitrary data that can be used for anything?

If I'm understanding that correctly, then I think that leaves us with a problem.  When considering computed fields, the docs say:

> The computed_field decorator can be used to include property or cached_property attributes when serializing a model or dataclass. **This can be useful for fields that are computed from other fields, ...**

What fields, you may ask?   That's exactly what I'd like to establish an answer for, and I'd like to do it in a way such that little pockets of the community don't solve with their own incompatible conventions.  

As a bit of background, I maintain libraries (like [psygnal](https://github.com/pyapp-kit/psygnal), [magicgui](https://github.com/pyapp-kit/magicgui), and others) that support pydantic models and provide runtime event-based GUIs.  If someone uses computed fields, it becomes very hard to know whether those values have become "stale" without knowing which fields those computed fields depend on.  

Can you help me understand what your current recommendation would be for that issue?  Is it indeed to come up with your own convention and encourage people to use it?

**Comment by sydney-runkle** [comment]
> I think it'd make sense to take this approach here and also sets us up well for future requests like this one.

I suppose more so what I meant here is that if we add a `metadata` attribute, folks can add all sorts of data to that depending on their varied needs, and that prevents us from the sort of bloating that we see with `ConfigDict` and `Field`, where we've catered to perhaps too many use cases.

I think you make a good point in general, I'm just hesitant to introduce a `dependencies` attribute. It feels to me like this implies either enforcement or some sort of reflection in the JSON schema, which is currently lacking.

I can see the value in having an endorsed way to communicate said dependencies, though. I'm torn.

@samuelcolvin is currently ooo, but should be back soon, and I can ask his opinion on this change when he returns? Then we could move forward from there?

**Comment by tlambert03** [comment]
Yep sounds good.  I understand the hesitation to be sure. 👍

**Comment by Viicos** [comment]
It is true indeed that with a metadata field instead of an explicit `dependencies` argument, there's no way to "enforce" the structure (at least from a documentation pespective). However, the primary (if not the only) feature `computed_field` fulfills is to be able to dump the property value (and thus having it included in JSON schema).

This _dependencies_ concept is also applicable for plain `@property`/`@functools.cached_property`. I could have the following plain dataclass defined:

```python
from dataclasses import dataclass

@dataclass
class A:
    a: int

    @property
    def prop(self) -> int:
        return self.a * 2
```

And a similar request to have dependencies specified would be valid as well.

---

You are currently supporting Pydantic models in your libraries but it could be very much possible to have someone requesting support for dataclasses, msgspec, (c)attrs, etc. One alternative way to specify these dependencies would be with a decorator:

```python
from pydantic import BaseModel

class Model(BaseModel):
    ...

    @property  # or `@computed_field`
    @depends_on("...")
    def prop(self): ...
```

That way you enforce yourself how dependencies should be specified, and it could be used in the future for other dataclass-like libraries.

**Comment by tlambert03** [comment]
Thanks @Viicos I agree on all cases.  I don't actually just support pydantic, I support arbitrary dataclass-likes via fieldz: https://github.com/pyapp-kit/fieldz

I recognize that this could be a general pattern, but so too is "computed_field" in general.  Pydantic didn't really need to add the computed_field decorator in the first place, simply supporting plain properties would have worked fine.

I am definitely always considering ways that I can do this stuff on my own, but inasmuch as a library like pydantic begins to formalize these concepts that I've generally made my own workarounds for, I'm interesting in discussing these formalizations in such a way that not everyone has to roll their own decorator solutions.  I get that it doesn't immediately solve the entire ecosystem problem, but given pydantics place in that ecosystem, it does seem like a good place to discuss these conventions in a way such that we can possibly all unify.  That's my interest here.

**Comment by tlambert03** [comment]
So, to perhaps summarize/clarify the state of things here, if the pydantic core devs wish to avoid getting involved in any way in a specification/recommendation for declaring relationships between a `computed_field` and other fields (which I can definitely understand), we can close this issue and encourage third party libraries to just create their own solutions

**Comment by sydney-runkle** [comment]
> I am definitely always considering ways that I can do this stuff on my own, but inasmuch as a library like pydantic begins to formalize these concepts that I've generally made my own workarounds for, I'm interesting in discussing these formalizations in such a way that not everyone has to roll their own decorator solutions. I get that it doesn't immediately solve the entire ecosystem problem, but given pydantics place in that ecosystem, it does seem like a good place to discuss these conventions in a way such that we can possibly all unify. That's my interest here.

This is such a great point, it's almost worth putting it up on our contributors page as a model open source perspective. We very much appreciate your encouraging discussions regarding such formalizations.

> So, to perhaps summarize/clarify the state of things here, if the pydantic core devs wish to avoid getting involved in any way in a specification/recommendation for declaring relationships between a computed_field and other fields (which I can definitely understand), we can close this issue and encourage third party libraries to just create their own solutions

We would like to avoid the `dependencies` argument for now, so I'm going to close this as not planned.

> edit: if a metadata field was added, I'd be somewhat satisfied even with just a note in the pydantic documentation "recommending" a format for including dependencies in the metadata. Mostly just want to avoid a bunch of roll-your-own solutions to indicating relationships between computed and other fields

@tlambert03, could you please open an issue with the `metadata` concept as a feature request? I'd love to move forward with that approach.

**Pull request closed without merging**